firefox-ios/CredentialProvider/CredentialProviderPresenter.swift

```diff
@@ -29,25 +29,37 @@ class CredentialProviderPresenter {
     }
 
     func credentialProvisionRequested(for credentialIdentity: ASPasswordCredentialIdentity) {
-        if self.profile.logins.reopenIfClosed() != nil {
-            cancel(with: .failed)
-        } else if let id = credentialIdentity.recordIdentifier {
-            profile.logins.getLogin(id: id, completionHandler: { [weak self] result in
-                switch result {
-                case .failure:
-                    self?.cancel(with: .failed)
-                case .success(let record):
-                    if let passwordCredential = record?.passwordCredential {
-                        self?.view?.extensionContext.completeRequest(
-                            withSelectedCredential: passwordCredential,
-                            completionHandler: nil
-                        )
-                    } else {
-                        self?.cancel(with: .userInteractionRequired)
+        let maxRetries = 3
+        var currentRetry = 0
+
+        func attemptProvision() {
+            if self.profile.logins.reopenIfClosed() != nil {
+                self.cancel(with: .failed)
+            } else if let id = credentialIdentity.recordIdentifier {
+                profile.logins.getLogin(id: id, completionHandler: { [weak self] result in
+                    switch result {
+                    case .failure:
+                        self?.cancel(with: .failed)
+                    case .success(let record):
+                        if let passwordCredential = record?.passwordCredential {
+                            self?.view?.extensionContext.completeRequest(
+                                withSelectedCredential: passwordCredential)
+                        } else {
+                            if currentRetry < maxRetries {
+                                currentRetry += 1
+                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+                                    attemptProvision()
```

Comment:
Just a thought but I would love if we had some telemetry to see how many times it fails and see if the timing and retries are enough or we need something more fancy exponential backoff.

Reply:
We don't have telemetry for this type of situations but I can add some logs.